File: Depth-First Search/Binary tree/Find paths/LET_no1008_Construct Binary Search Tree from Preorder Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def bstFromPreorder(self, preorder: List[int]) -> Optional[TreeNode]:
        return self.helper(preorder[::-1], float('inf'))
    
    def helper(self, nums, bound):
        if not nums or nums[-1] > bound:
            return None
        
        root = TreeNode(nums.pop())
        root.left = self.helper(nums, root.val)
        root.right = self.helper(nums, bound)
        
        return root

File: Depth-First Search/Binary tree/Find paths/LET_no1008_Construct Binary Search Tree from Preorder Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* bstFromPreorder(vector<int>& preorder) {
        int i = 0;
        return helper(preorder, i, INT_MAX);
    }
    
    TreeNode* helper(vector<int>& nums, int& i, int bound) {
        if (i == nums.size() || nums[i] > bound)
            return nullptr;
        
        TreeNode* root = new TreeNode(nums[i++]);
        root->left = helper(nums, i, root->val);
        root->right = helper(nums, i, bound);
        
        return root;
    }
};